src/client/Microsoft.Identity.Client/OAuth2/MsalTokenResponse.cs

```diff
@@ -146,6 +146,57 @@ internal static MsalTokenResponse CreateFromiOSBrokerResponse(Dictionary<string,
             return response;
         }
 
+        internal static MsalTokenResponse CreateFromAppProviderResponse(ExternalTokenResult externalTokenResponse)
+        {
+            ValidateExternalTokenResult(externalTokenResponse);
+
+            var response = new MsalTokenResponse
+            {
+                AccessToken = externalTokenResponse.RawAccessToken,
+                RefreshToken = null,
+                IdToken = null,
+                TokenType = BrokerResponseConst.Bearer,
+                ExpiresIn = externalTokenResponse.ExpiresInSeconds,
+                ClientInfo = null,
+                TokenSource = TokenSource.IdentityProvider,
+                TenantId = externalTokenResponse.TenantId
+            };
+
+            if (externalTokenResponse.RefreshInSeconds.HasValue)
+            {
+                response.RefreshIn = externalTokenResponse.RefreshInSeconds.Value;
+            }
+            else
+            {
+                response.RefreshIn = response.ExpiresIn;
+            }
+
+            return response;
+        }
+
+        private static void ValidateExternalTokenResult(ExternalTokenResult externalTokenResult)
+        {
+            if (string.IsNullOrEmpty(externalTokenResult.RawAccessToken))
```

Comment:
Do we need a check that if `RefreshInSeconds.HasValue`, `RefreshInSeconds.HasValue >= 0`? Unless we check that when we do refresh.

Reply:
It is checked when we do refresh.